### ISSUE TITLE
The performance gets worse when I use a bigger model

### ISSUE THREAD

**Issue by Marcovaldong** [HawkAaron/RNN-Transducer#14]
@HawkAaron I tried to train the transducer on my own dataset. When I increase the hidden_size from 250 to 512 and increase the num_layers from 3 to 5, the performance became worse, why? The bigger model should be over-fitting more easily, isn't it?

**Comment by HawkAaron** [comment]
What's your performance measurement? And how did you compare the performance ?

**Comment by Marcovaldong** [comment]
I am training the transducer on a Chinese dataset, the label is pinyin, the measurement is WER/CER. My goal is to measure the performance of transducer on a big dataset (I should train it on several thousand hours audio data). To verify there is no problem in my code, I train it on a little dataset: the train set contains 2000 samples, the val set and the test set are the same one, they contains 200 samples choosed randomly from the train set. 

I changed the model structure. The encoder network contains an unidirectional multi-layer lstm (which num_layer is num_layers1), a bidirectional lstm and a fc layer. The decoder network contains a single or a multi-layer unidirectional lstm (which num_layer is num_layers2). The joint part is same with your code. 
The hidden_size of all lstm layers are equal. So I have three super parameters: num_layers1, num_layers2 and num_hidden.

I changed super parameters and trained different transducers. I found that bigger model has worse performance (when the super-params is 5, 1, 512, the loss is continuing decreased, however the model cannot even translate any audio.) and the smaller model fit the dataset more quickly.  (the model with super-params (2, 1, 250) is better than the model with super-params (3, 1, 250))

Is there some code are inferenced by these super-params?

**Comment by HawkAaron** [comment]
Maybe its not enough to train your model on only 2000 samples. The model may overfit on the little dataset, but cannot translate anything.  

There are some suggestions:
1. Init transcription network with CTC trained encoder,
2. Init prediction network with CE trained LM,
3. Use Adam or Adadelta to train your model.

**Comment by Marcovaldong** [comment]
@HawkAaron Thanks for your reply. I always use Adam as the optimizer. I don't know how to upload pretrained params for some layers of the model, I'm learning it. 

The result of my experiments shows me that the loss is proportional to the error rate: the loss higher, the CER higher. The loss lower, the CER lower. The bigger model get higher loss and higher CER (even great than 100%), the smaller model get lower loss and lower CER.

**Comment by HawkAaron** [comment]
For such small data (2000samples), the model can easily overfit, generally, the loss should be less than 10 after several epochs, if not, you may check the output label, especially the blank label (default 0, then label index start from 1).

**Comment by Marcovaldong** [comment]
Yes, the loss is less than 10, however the CER is high. The label should be correct.

You means that since the model get low loss on trainset, but it cannot translate the trainset, is it? That means the trainset is too small for the model to get a good performance, is it?

**Comment by HawkAaron** [comment]
Maybe you could try CTC model first.

**Comment by Marcovaldong** [comment]
I am training CTC model with the same method, there may be the same situation. Otherwise, I am training CTC model implemented by PyTorch. Tomorrow, I will get the result.

**Comment by Marcovaldong** [comment]
Is there a balance between the encoder network and the prediction network? I checked three papers just know. In [1801.00841](https://arxiv.org/abs/1801.00841), the encoder network contains 12 LSTM layers and a time convolution, the decoder contains 2 lstm layers. In [1811.06621](https://arxiv.org/abs/1811.06621), there are 8 lstm layers in encoder and 2 lstm layers in prediction part. In [1811.05097](https://arxiv.org/abs/1811.05097), there are 5 lstm layers and an intermediate supervision in encoder and 2 lstm layers in prediction part. 

Is there a balance between the encoder network and the prediction network? Two strong prediction part will inference the performance.  However, the purpose of the prediction part is to introduced the language model in the model. I'm confused.

**Comment by HawkAaron** [comment]
Well, its almost the same with attention based encoder-decoder, the encoder usually use deeper lstms, like 5 layers blstm with 512 hidden dim in each direction, the decoder usually use 2 layers lstm with 1024 hidden dim. 



发自我的小米手机在 marcovaldo <notifications@github.com>，2019年5月9日 15:10写道：Is there a balance between the encoder network and the prediction network? I checked three papers just know. In 1801.00841, the encoder network contains 12 LSTM layers and a time convolution, the decoder contains 2 lstm layers. In 1811.06621, there are 8 lstm layers in encoder and 2 lstm layers in prediction part. In 1811.05097, there are 5 lstm layers and an intermediate supervision in encoder and 2 lstm layers in prediction part.
Is there a balance between the encoder network and the prediction network? Two strong prediction part will inference the performance.  However, the purpose of the prediction part is to introduced the language model in the model. I'm confused.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/HawkAaron/RNN-Transducer/issues/14#issuecomment-490772242",
"url": "https://github.com/HawkAaron/RNN-Transducer/issues/14#issuecomment-490772242",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by Marcovaldong** [comment]
I did another experiment this afternoon. The train set contains 2000 samples which are the same sample (I would change this audio in different experiments). The CTC model can decode the audio correctly after 2 epochs. However, for some samples, the transducer need to be trained for more than 50 epochs. 

It looks like the language model part inferences the performance of the total model. Maybe the pretrained language model will accelerates the training procedure.
